### PR TITLE
Setting the payment object again

### PR DESCRIPTION
**Pull request by johnchque**
This should work, setting the payment object again.

### CODE REVIEW

**Comment by bartfeenstra** [comment]
Thanks for this PR! This is a real problem, but a hard one to track. Would it perhaps be better if we do this in the class' constructor for all payment-aware plugins (payment method, statuses, line items, and type)?

**Comment by Berdir** [comment]
I think the problem is that the payment objects can get out of sync, e.g. due to serialization/unserialization and you end up with multiple instances. Not sure if setting it in the constructor would help, I believe that's basically already happening and not enough. Or maybe I misunderstand you.

This is certainly a hacky solution and I'm open for alternatives.

**Comment by bartfeenstra** [comment]
It doesn't happen in `\Drupal\payment\Entity\Payment::__construct()` yet. It's effectively the only place where we can reliably catch occurrences of this problem, I think. When setting instances on plugin fields, the correct payment is set automatically, but when loading entities all that code is bypassed.

**Comment by Berdir** [comment]
Had another look at this, and no, it IMHO has nothing to do with __construct(). This is about serialized entities, __construct is not called for those.

There are also already multiple attempts and workarounds in place. There is \Drupal\payment\Entity\Payment::getTranslatedField(), which tries to do this, but looking at \Drupal\Core\Entity\ContentEntityBase::get() only calls it if the field isn't instantiated yet. I assume what happens is that we call $payment->setPaymentType(), with a payment type from the plugin selector that references an old, different payment instance.

That's all quite weird, maybe a better workaround is to call setPayment() in setPaymentMethod()?

That said, it is worth mentioning that \Drupal\payment\Entity\Payment::setPaymentStatus() has exactly the same workaround.

**Comment by bartfeenstra** [comment]
You're right. This PR makes sense.

**Comment by bartfeenstra** [comment]
Let's keep this call to `::setPayment()`, but move it to `::setPaymentMethod()` instead, so we benefit in all situations where we call `::getPaymentMethod()`.

File: src/Entity/Payment.php [hunk 1]
```diff
@@ -281,7 +281,10 @@ public function setPaymentMethod(PluginPaymentMethodInterface $payment_method) {
     /** @var \Drupal\plugin\Plugin\Field\FieldType\PluginCollectionItemInterface $field_item */
     $this->get('payment_method')->applyDefaultValue();
     $this->get('payment_method')->appendItem($payment_method);
-
+    // The payment method might have been unserialized with an old payment
+    // object, trying to save that as new will result in exceptions. Set the
+    // current object again.
+    $this->getPaymentMethod()->setPayment($this);
```
Comment (Berdir) [inline src/Entity/Payment.php]: Call $payment_method->setPayment($this) instead, possibly before appending it, shouldn't make a difference but makes a bit more sense when reading.

File: src/Tests/Entity/PaymentUnitTest.php [hunk 2]
```diff
@@ -174,6 +174,7 @@ protected function testGetPaymentMethod() {
     $payment_method = Payment::methodManager()->createInstance('payment_basic');
     $this->payment->setPaymentMethod($payment_method);
     $this->assertTrue(spl_object_hash($this->payment->getPaymentMethod()), spl_object_hash($this->payment));
```
Comment (Berdir) [inline src/Tests/Entity/PaymentUnitTest.php]: Xano agreed that this assert doesn't make sense, lets fix it while we touch the code here.

**Comment by Berdir** [comment]
The last commit looks pretty fine but something is not right with the commits here, maybe the most recent commits weren't pushed to github?

**Comment by bartfeenstra** [comment]
@Berdir Both Github and d.o repos are up to date. I assume 8.x-2.x has not been merged into the feature branch for a while.

**Comment by bartfeenstra** [comment]
I committed this PR's patch directly at https://www.drupal.org/node/2630804, so I could give you drupal.org credit as well. Thanks!

**Comment by johnchque** [comment]
Thank you so much.

**Pull request closed without merging**